### PR TITLE
Issue 22791 render pagee with variant

### LINKED ISSUE #22790: Create default variant
We need to create a `default variant`, so we need to:

**Create a new TaskUpgrade to:**

- Create a `default variant` in the variant table
- Add a new variant field to "contentlet_version_info", the default value should point to the `default variant`

### LINKED ISSUE #22791: Render page using variants
We need to be able to use variant when we want to render a page, so we must redefine the fallback logic to follow these steps:

- First we should look for the current language and current variant.
- If it does not exist, we should look for variant and default language.
- If it does not exists, we should look for default variant and language.
- If it does not exists, we should look for default variant and default language.

Also we need to handle a current variant as we handle the current language right now, it means:

- When you send a request parameter with the current variant then it overwrite the one into sessio.
- When the variant is not send in the request then we take it from the session.
- When it does not exists neither into session nor request then we take the default one.

We need to update the follows codes to render by current variants and current languages:

https://github.com/dotCMS/core/blob/99c17c27870101a1f2fcf8fa0575251d281b10e2/dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/HTMLPageAssetRenderedAPIImpl.java#L374

https://github.com/dotCMS/core/blob/a3cda524b8220ca5801bfd8b402ed274b6cb8145/dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/ContentletDetail.java#L35

https://github.com/dotCMS/core/blob/b6bb9d3a4f3eb4507de09088b99427fb67b8932f/dotCMS/src/main/java/com/dotcms/rendering/velocity/services/ContentletLoader.java#L570

### PR DESCRIPTION
**Pull request by freddyucv**
Work in progress

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageLoader.java [hunk 1]
```diff
@@ -201,9 +201,9 @@ private void addWidgetPreExecuteCodeAndPageInfo(final HTMLPageAsset htmlPage, fi
     public InputStream writeObject(final VelocityResourceKey key) throws DotDataException, DotSecurityException {
 
         HTMLPageAsset page = APILocator.getHTMLPageAssetAPI()
-            .fromContentlet(APILocator.getContentletAPI()
-                .findContentletByIdentifier(key.id1, key.mode.showLive, Long.parseLong(key.language), sysUser(), true));
-
+                .fromContentlet(APILocator.getContentletAPI()
+                .findContentletByIdentifier(key.id1, key.mode.showLive, Long.parseLong(key.language),
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageLoader.java:205]: usually is better to use ConversionUtil.toLong(key.language, ()-> ApiLocator.getLanguageAPI().getDefaultLanguage().getId())

File: dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPI.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * @param identifier {@link com.dotcms.content.model.Contentlet}'s identifier
+     * @param pageMode page mode to render
+     * @param user to check {@link com.dotmarketing.beans.Permission}
+     * @return
+     */
+    RenderContext getRenderContext(final long tryingLang, final String identifier,
+            final PageMode pageMode, final User user);
+
+    class RenderContext {
+        private String currentVariantKey;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPI.java:32]: set to final

File: dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPIImpl.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                        .get(requestParameter);
+                return byName.isPresent() ? byName.get().name() : VariantAPI.DEFAULT_VARIANT.name();
+            } catch (DotDataException e) {
+                Logger.error(VariantWebAPIImpl.class,
+                        String.format("It is not possible get variant y name %s: %s",
+                                requestParameter, e.getMessage()));
+                return VariantAPI.DEFAULT_VARIANT.name();
+            }
+        } else {
+            return VariantAPI.DEFAULT_VARIANT.name();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPIImpl.java:62]: you can do:
```
if (UtilMethods.isSet(requestParameter)) {
            try {
                final Optional<Variant> byName = APILocator.getVariantAPI()
                        .get(requestParameter);
                return byName.isPresent() ? byName.get().name() : VariantAPI.DEFAULT_VARIANT.name();
            } catch (DotDataException e) {
                Logger.error(VariantWebAPIImpl.class,
                        String.format("It is not possible get variant y name %s: %s",
                                requestParameter, e.getMessage()));
         
            }
        }
            return VariantAPI.DEFAULT_VARIANT.name();
```

Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPIImpl.java:62]: really I like the else there

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageLoader.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageLoader.java]: done https://github.com/dotCMS/core/pull/23022/commits/4d91d0bf79d06f7b4cca64afb251f042ac7520a1#diff-d285e87fa41479227733496517a21cc9f98fbe838f07268bfafec8fc298ddf61R206

File: dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPI.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPI.java]: done https://github.com/dotCMS/core/pull/23022/commits/4d91d0bf79d06f7b4cca64afb251f042ac7520a1#diff-45dbdd64c70c9bbcc5509ea0609065408ee20034d258b0ffa416e73c6976f0f5R32

File: dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPIImpl.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+import javax.servlet.http.HttpServletRequest;
+import org.apache.velocity.exception.ResourceNotFoundException;
+
+/**
+ * Default implementation for {@link VariantWebAPI}
+ */
+public class VariantWebAPIImpl implements VariantWebAPI{
+
+    /**
+     * Return the current version following this steps:
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/variant/business/web/VariantWebAPIImpl.java:29]: current Variant*

**Review by dsilvam — APPROVED** [review]

**Review by jdotcms — APPROVED** [review]

**Pull request merged**